File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.js
function rotate(gear, direction) {
  let rotatedGear = Array(8).fill(0);
  if (direction === 1) {
    // 시계방향 회전
    for (let i = 0; i < 7; i++) {
      rotatedGear[i + 1] = gear[i];
    }
    rotatedGear[0] = gear[7];
  } else {
    // 반시계방향 회전
    for (let i = 1; i < 8; i++) {
      rotatedGear[i - 1] = gear[i];
    }
    rotatedGear[7] = gear[0];
  }
  return rotatedGear;
}

function moveGear(gears, num, direction, visited) {
  if (visited[num]) return;
  visited[num] = true;

  if (num > 0 && gears[num][6] !== gears[num - 1][2]) {
    moveGear(gears, num - 1, -direction, visited);
  }

  if (num < 3 && gears[num][2] !== gears[num + 1][6]) {
    moveGear(gears, num + 1, -direction, visited);
  }

  gears[num] = rotate(gears[num], direction);
}

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./14891.txt";
let input = fs.readFileSync(filePath).toString().trim().split("\n");

const arr = input.slice(0, 4).map((l) => l.trim().split("").map(Number));
const K = parseInt(input[4]);
const orders = input
  .slice(5, 5 + K)
  .map((o) => o.trim().split(" ").map(Number));

let temp = arr.map((row) => [...row]);

orders.forEach((order) => {
  let visited = Array(4).fill(false);
  const [num, direction] = order;
  moveGear(temp, num - 1, direction, visited);
});

let ans = temp.reduce((acc, val, idx) => acc + val[0] * Math.pow(2, idx), 0);
console.log(ans);

File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.py
import sys
from collections import deque

input = sys.stdin.readline


def rotate_gears(gears, num, direction):
    """주어진 톱니(num, 방향) 기준으로 좌우 전파를 체크하고 회전"""
    need_change = [0, 0, 0, 0]
    need_change[num] = direction

    # 왼쪽 전파
    d = direction
    for i in range(num - 1, -1, -1):
        if gears[i][2] != gears[i + 1][6]:
            d *= -1
            need_change[i] = d
        else:
            break

    # 오른쪽 전파
    d = direction
    for i in range(num + 1, 4):
        if gears[i - 1][2] != gears[i][6]:
            d *= -1
            need_change[i] = d
        else:
            break

    # 실제 회전
    for i in range(4):
        if need_change[i] != 0:
            gears[i].rotate(need_change[i])


def calc_score(gears):
    """톱니바퀴의 최종 점수를 계산"""
    score = 0
    for i in range(4):
        if gears[i][0] == "1":  # 12시 방향
            score += 2**i  # 2^i
    return score


# --- 실행 부분 ---
gears = [deque(input().strip()) for _ in range(4)]
K = int(input())

for _ in range(K):
    num, direction = map(int, input().split())
    rotate_gears(gears, num - 1, direction)

print(calc_score(gears))
